modules/light-clients/10-attestations/client_state.go

```diff
@@ -0,0 +1,247 @@
+func (cs *ClientState) verifyNonMembership(
+	clientStore storetypes.KVStore,
+	cdc codec.BinaryCodec,
+	height exported.Height,
+	proof []byte,
+	path exported.Path,
+) error {
+	if cs.IsFrozen {
+		return ErrClientFrozen
+	}
+
+	if path == nil || path.Empty() {
+		return errorsmod.Wrap(ErrInvalidPath, "path cannot be empty")
+	}
+
+	if _, found := getConsensusState(clientStore, cdc, height); !found {
+		return errorsmod.Wrapf(clienttypes.ErrConsensusStateNotFound, "consensus state not found for height %s", height)
+	}
+
+	var attestationProof AttestationProof
+	if err := cdc.Unmarshal(proof, &attestationProof); err != nil {
+		return errorsmod.Wrapf(ErrInvalidAttestationProof, "failed to unmarshal proof: %v", err)
+	}
+
+	if err := cs.verifySignatures(&attestationProof); err != nil {
+		return err
+	}
+
+	packetAttestation, err := ABIDecodePacketAttestation(attestationProof.AttestationData)
+	if err != nil {
+		return errorsmod.Wrapf(ErrInvalidAttestationData, "failed to ABI decode attestation data: %v", err)
+	}
+
+	if packetAttestation.Height != height.GetRevisionHeight() {
+		return errorsmod.Wrapf(ErrInvalidHeight, "height mismatch: expected %d, got %d", height.GetRevisionHeight(), packetAttestation.Height)
+	}
+
+	if len(packetAttestation.Packets) == 0 {
+		return errorsmod.Wrap(ErrInvalidAttestationData, "packets cannot be empty")
+	}
+
+	commitmentPath, err := canonicalizePath(path)
+	if err != nil {
+		return err
+	}
+
+	foundMatchingPath := false
+	allZeroCommitments := true
+	for _, packet := range packetAttestation.Packets {
+		if bytes.Equal(packet.Path, commitmentPath) {
+			foundMatchingPath = true
+			if len(packet.Commitment) != 32 || !bytes.Equal(packet.Commitment, zeroCommitment) {
+				allZeroCommitments = false
+			}
+		}
+	}
+
+	if !foundMatchingPath {
+		return ErrNotMember
+	}
+
+	if !allZeroCommitments {
+		return ErrNonMembershipFailed
+	}
+
+	return nil
+}
+
+func canonicalizePath(path exported.Path) ([]byte, error) {
+	switch p := path.(type) {
+	case commitmenttypesv2.MerklePath:
+		flattened := bytes.Join(p.GetKeyPath(), []byte("/"))
+		return normalizePathBytes(flattened), nil
+	case *commitmenttypesv2.MerklePath:
+		flattened := bytes.Join(p.GetKeyPath(), []byte("/"))
+		return normalizePathBytes(flattened), nil
+	case interface{ GetKeyPath() [][]byte }:
+		flattened := bytes.Join(p.GetKeyPath(), []byte("/"))
+		return normalizePathBytes(flattened), nil
+	case interface{ Bytes() []byte }:
+		return normalizePathBytes(p.Bytes()), nil
+	default:
+		if stringer, ok := path.(fmt.Stringer); ok {
+			return normalizePathBytes([]byte(stringer.String())), nil
+		}
+	}
+
+	return nil, errorsmod.Wrapf(ErrInvalidPath, "unsupported path type %T", path)
+}
+
+func normalizePathBytes(raw []byte) []byte {
+	if len(raw) == 32 {
+		return raw
+	}
+	return crypto.Keccak256(raw)
+}
+
+func normalizeValueBytes(raw []byte) []byte {
+	if len(raw) == 32 {
+		return raw
+	}
+	return crypto.Keccak256(raw)
+}
```

Comment:
Is this really true? I thought the attestor doesn't attest to anything except the packet commitment. This function shouldn't really exist. We can simply check that `len(raw) == 32`, else error